tribits/ci_support/CreateIssueTrackerFromCDashQuery.py

```diff
@@ -0,0 +1,271 @@
+
+g_pp = pprint.PrettyPrinter(indent=2)
+
+
+# Class that is used to create issue trackers for nonpassing tests downloaded
+# from a CDash queryTests.php query.
+#
+# To make this general, the user has to create a class object that is passed
+# in as the argument 'issueTrackerFormatter' in this class that supports
+# the following functions.
+#
+# issueTrackerFormatter.acceptIssueTrackerData(issueTrackerData):
+#
+# This function is given data about the nonpasing tests through an object of
+# type 'IssueTrackerData' as the argument 'issueTrackerData'.  This data is
+# used to create the issue issue tracker text.
+# issueTrackerFormatter.getIssueTrackerText():
+#
+# This function is expected to return the created text for the issue tracker
+# given the data provided in the above 'acceptIssueTrackerData()' function.
+# The format of this text can be anything such as Markdown, RestructredText,
+# JIRA Issue Markup, etc.  The only assumption about formatting is in the HTML
+# table passed in the arugment 'issueTrackerData.testHistoryHtmlTableText'
+# 'acceptIssueTrackerData()' function.
+#
+# The object 'issueTrackerFormatter' can format the issue tracker text anyway
+# it wants.  The object 'issueTrackerFormatter' is stateless w.r.t. this class
+# CreateIssueTrackerFromCDashQueryDriver
+#
+class CreateIssueTrackerFromCDashQueryDriver:
+
+
+  def __init__(self, issueTrackerFormatter, cdashProjectStartTimeUtc=None,
+      usageHelp="", issueTrackerUrlTemplate="", issueTrackerTemplate="",
+    ):
+    self.issueTrackerFormatter = issueTrackerFormatter
+    self.cdashProjectStartTimeUtc = cdashProjectStartTimeUtc
+    self.usageHelp = usageHelp
+    self.issueTrackerUrlTemplate = issueTrackerUrlTemplate
+    self.issueTrackerTemplate = issueTrackerTemplate
+    self.options = None
+
+
+  def runDriver(self):
+
+    print("\n***")
+    print("*** Getting data to create a new issue tracker")
+    print("***\n")
+
+    self.getCmndLineOptions()
+
+    print("Downloading full list of nonpassing tests from CDash URL:\n")
+    print("   "+self.options.nonpassingTestsUrl+"\n")
+
+    nonpassingTestsLOD = self.downloadNonpassingTestsData()
+    print("\nTotal number of nonpassing tests over all days = "\
+      +str(len(nonpassingTestsLOD)))
+
+    uniqNonpassingTestsLOD = getUniqueTestsListOfDicts(nonpassingTestsLOD)
+    print("\nTotal number of unique nonpassing test/build pairs over all days = "\
+          +str(len(uniqNonpassingTestsLOD)))
+
+    (testnameList, buildnameList) = getTestnameAndBuildnameLists(uniqNonpassingTestsLOD)
+    print("\nNumber of test names = "+str(len(testnameList)))
+    print("\nNumber of build names = "+str(len(buildnameList)))
+
+    testingDayStartNonpassingDate = getTestingDayStartNonpassingDate(
+      nonpassingTestsLOD, self.options.cdashProjectStartTimeUtc)
+    #print("testingDayStartNonpassingDate = "+testingDayStartNonpassingDate)
+
+    testHistoryHtmlTableText = ""  # ToDo: Implement!
```

Comment:
Should this be implemented yet, or not quite?

Reply:
It would be useful to have this test history table but I wanted an initial version of the tool that I could use more quickly  I wanted to put in a placeholder for this table since it will take a little time to implement and test.  But it would not be that hard to do and it would really help provide context for randomly failing tests.